### DISCUSSION TITLE
Unable to distinguish between unset and `None` values in `IntegrityError.params`

### DISCUSSION THREAD

**Discussion by auxsvr** [sqlalchemy/sqlalchemy#10768] [Usage Questions]
Trying to handle some uniqueness constraint violations automatically, it seems that there is no way to distinguish between unset and `None` values in `IntegrityError.params`. Something that might be related is that the `insert` statement produced by the ORM sets all unset columns to `None`. Do you think it would be more appropriate to simply omit them? Is there a way to find the unset columns in `IntegrityError.params`?

**Comment by zzzeek** [comment]
The ORM omits columns from INSERT statements that have client or server side defaults configured on them (such as FetchedValue, server_default=x, etc), however for columns that have no defaults, it passes in None which is to allow it to be able to more efficiently batch common INSERT statements together.

if you have IntegrityError alone, there's no way to determine which columns were unset vs. None, you'd need to look at the ORM mapped object itself and I would advise setting up validation at the ORM layer if you are trying to be precise about this.      You might also try configuring FetchedValue defaults on particular columns that you are seeing as sensitive to being present or not in an INSERT statement.

**Reply by auxsvr** [reply]
Thank you for the prompt response!

It seems that there is a way to resolve uniqueness constraints automatically and reliably in case the uniqueness constraint columns comprise a secondary or alternate key. Some code that seems to work after a few quick tests (this is in the `commit` method of a subclass of `AsyncSession`):
```python
        pending = self.dirty | self.new
        """Objects that will produce update or insert statements."""
        try:
            await commit()
        except sqlalchemy.exc.IntegrityError as e:
            if not (resolve_conflicts and isinstance(e.orig, (psycopg.errors.UniqueViolation,))):
                raise
            logger.info(f"Attempting to resolve conflict {e.args[0]!r}...")
            diagnostics = e.orig.diag
            if class_ := get_model_from_table_name(diagnostics.schema_name, diagnostics.table_name):
                # We use the values from the mapped object instead of the exception, as they allow
                # distinguishing between None and unset columns.
                models = {m: model2dict(m) for m in pending if isinstance(m, class_)}
                await self.rollback()  # This must be done before accessing the session again.
                constraints = await get_constraint_names(
                    diagnostics.schema_name, diagnostics.table_name, False
                )
                # Get the object whose columns that are in the conflicting constraint have the
                # conflicting values. This works reliably, because the uniqueness constraint
                # defines a key, an alternate or secondary key.
                model = None
                for m in models:
                    constraint_values = {
                        col: models[m][col] for col in constraints[diagnostics.constraint_name]
                    }
                    if all(
                        v is not None and v == e.params[k] for k, v in constraint_values.items()
                    ):
                        model = m
                        break
                if model is None:
                    logger.warn(
                        "Unable to find an object matching the secondary key and resolve the conflict."
                    )
                    raise
                non_constraint_values = {
                    k: v for k, v in models[model].items() if k not in constraint_values
                }
                if not non_constraint_values:
                    persistent = (
                        await self.execute(
                            select(class_)
                            .where(
                                and_(
                                    *(
                                        getattr(class_.__table__.c, k) == v
                                        for k, v in constraint_values.items()
                                    )
                                )
                            )
                        )
                    ).scalar_one()
                    logger.info("The conflicting object already exists in the database.")
                else:
                    persistent = await self.scalar(
                        insert(class_)
                        .values(**models[model])
                        .on_conflict_do_update(
                            constraint=diagnostics.constraint_name, set_=models[model]
                        )
                        .returning(class_)
                        .execution_options(populate_existing=True)
                    )
                    await commit()
                # Update any mapped object identified by the secondary key.
                pk_name = (
                    sqlalchemy.inspect(class_)
                    .get_property_by_column(model.__table__.primary_key.columns[0])
                    .key
                )
                # Update the object's primary key, so that the mapper identifies it correctly.
                setattr(model, pk_name, getattr(persistent, pk_name))
                logger.info(
                    f"Conflict on constraint {diagnostics.constraint_name} resolved automatically!"
                )
            else:
                logger.warn("Failed to resolve conflict...")
                raise
```
Is there any obvious (or even not so obvious!) reason this would fail in practice? Is there an easier way to do this?

**Reply by zzzeek** [reply]
I usually would avoid trying to catch IntegrityErrors and "automatically" resolve them generically like that.  If I have a part of my program where I'm inserting a row and I think it might have an integrity error, I would write code specific to that INSERT statement to handle that contingency, either by using an UPSERT instead to start with, or if catching an integrity error, by explicitly converting to an UPDATE.     The approach above looks very complicated and it seems that it would be better handled by specific use cases that expect row conflicts to include explicit directives on how conflicts should be resolved.

I would also ensure the entire handling of the condition is in a single transaction, which for PostgreSQL would mean I'd use a savepoint if I expect to catch an integrity error and then continue. Otherwise, if this IntegrityError is caught as part of a series of statements within a larger transaction, it doesn't roll back all the other work that was done and just throw it away.

**Reply by auxsvr** [reply]
Do you think it would make sense to update the mapper to handle a secondary key as an alternative for the primary key in such cases? In theory, it should handle them correctly, although it seems that the code would be fairly complicated.

**Reply by zzzeek** [reply]
im not sure what you mean.  do you mean should we integrate "upsert" into the ORM unit of work process?    answer there is, it would be an enormous and complicated change where it would be extremely critical that the chosen public API and behavioral contract is nailed perfectly, and I'm not sure what API could be chosen for that which will not introduce more problems than it attempts to solve.

**Reply by auxsvr** [reply]
Yes, thank you. I'll see how some ideas I have work in practice and determine if this is worth pursuing.
